doc: Use dt binding title property for board supported hardware table

asmellby opened the pull request:
The `title` property was added to dt bindings in #86533. Expose it on the Node class similar to how `description` is propagated, and make use of it for the `zephyr:board-supported-hw` directive in the documentation.

Leverage `title` to improve how supported hardware peripherals appear in board documentation for Silicon Labs boards.

asmellby committed:
doc: _scripts: Expose title from dt binding in board catalog

Make the recently added `title` attribute in dt bindings available
in the board catalog used by the board supported hardware directive.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
edtlib: Expose binding title in node class

Make the binding title available from the node the same way
the binding description is propagated.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
dts: bindings: silabs: Clean up descriptions, add titles

Clean up dt binding descriptions and introduce titles where
needed to make board documentation pages look nice. The supported
hardware table on board documentation pages sources its data
from dt bindings, and needs succinct titles.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

kartben reviewed (commented):
This is awesome - thanks for doing this @asmellby, you beat me to it :)

One thing that I was asking myself when support for this new property was added is that it feels like we need some guidelines as to what makes a "good" title. Thinking out loud here, there's two things that I'm undecided about:

- Should the vendor even appear in the title? We already have that info in the binding's compatible so we could use that if/where needed. I think a product line can be useful though.

- How about acronyms? Don't get me wrong, I do want titles to be short and concise, but OTH I wouldn't be surprised if most people, just like me, have no idea what "HFRCODPLL" stands for :) 

So basically, should `Silicon Labs HFRCODPL` be `Series 2 HFRCODPLL (High-Frequency RC Oscillator with Digital Phase-Locked Loop)`? FWIW, if we are really clear in our guidelines as to what's allowed in a title -- and possibly implement associated automated checks -- we could even have doc generation automatically "compact" the acronym so as it shows up as a tooltip thingy just like other acronyms like ADC etc...

And on the topic of tooltip, I know the compatible string links to the bindings' doc, but I guess we could also have a tooltip with the full description when one hovers over the compatible string (and that enhancement would benefit all users of `:dtcompatible:` role...)

Anyhow, thanks again, really appreciate it!

asmellby committed:
doc: _extensions: zephyr: Use binding title in supported hw table

Use the title property from bindings in the supported hardware table
if available.

Create <abbr> elements for acronyms and abbreviations in titles.
Abbreviations are not created when the description field is used as
a title, since existing text in bindings was not created with this
in mind, and parentheses are used in the text in ways that are
incompatible with the auto-abbreviating logic.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby commented:
I agree with your thoughts -- repeating the vendor name for every on-chip peripheral gets very repetitive, but family/product line makes sense. I really like your idea to parse out the acronyms/abbreviations -- the reason I didn't expand them originally was that things got noisy when every cell in the table spanned multiple lines, but with an `<abbr>` element we get the best of both worlds. I took the liberty of implementing it in this PR.

kartben commented on doc/_extensions/zephyr/domain/__init__.py:
```diff
@@ -929,7 +929,7 @@ def feature_sort_key(feature):
                     desc_entry = nodes.entry(classes=["description"])
                     desc_para = nodes.paragraph(classes=["status"])
-                    desc_para += nodes.Text(value["description"])
+                    desc_para += parse_text_with_acronyms(value["description"])
```
Nice!..... But... the reason I was mentioning we'd need really strict rules is that it really only works with your new perfectly formed titles, and doesn't work so great if the description is a legacy "freeform" description (think "GPIO pins exposed on Arduino Uno (R3) headers", "ESP32 Two-Wire Automotive Interface (TWAI)", ...)
There's many ways to fix this, but not sure what's the best one short-term. Be creative? :)

In any case, if we go for this automagic abbreviation thing (which might still be a bad idea, after all, since e.g it kind of prevents people from using brackets for anything else), we do need to document somewhere how it works so that people know how to write a "proper" title

asmellby replied:
We could also make the regex more restrictive -- it currently matches on any word (upper- or lowercase) followed by parantheses. We might only want auto-acronyms for words that are fully upper case.

kartben replied:
> We could also make the regex more restrictive -- it currently matches on any word (upper- or lowercase) followed by parantheses. We might only want auto-acronyms for words that are fully upper case.

Sounds like a really good idea to me!

asmellby replied:
Restricted the regex to uppercase+digits only. A quick grep of the `dts/bindings/` directory in the tree finds the following 25 false positives that would be wrongly abbreviated:

* `adi,ad5691`: AD5691 (12-bit)
* `adi,ad5692`: AD5692 (14-bit)
* `adi,ad5693`: AD5693 (16-bit)
* `ambiq,spi-bleif`: SOC (e.g. Apollo3 Blue)
* `atmel,24mac402`: AT24 (or compatible)
* `atmel,at45`: AT45 (or compatible)
* `atmel,sam-hsmci`: ATMEL (Microchip)
* `atmel,sam-i2c-twi`: I2C (TWI)
* `atmel,sam-i2c-twihs`: I2C (TWIHS)
* `atmel,sam-i2c-twim`: I2C (TWIM)
* `atmel,sam-usbc`: USB (USBC)
* `atmel,sam-usbhs`: USB (USBHS)
* `atmel,sam0-id`: ID (serial number)
* `nxp,iap-fmc11`, `nxp,iap-fmc54`, `nxp,iap-fmc55`, `nxp,iap-fmc553`: NXP (In-Application Programming)
* `smsc,lan91c111-mdio`, `smsc,lan91c111`, `smsc,lan9220`: SMSC (now Microchip)
* `ti,tlc5971`: TLC5971 (and compatible)
* `ti,tps382x`: IC (TPS382x)
* `ws2812`: WS2812 (and compatible)
* `zephyr,bt-hci-spi-slave`: SPI (HCI SPI)
* `zephyr,nus-uart`: NUS (Bluetooth LE)

What do you think we should do with these? Let them get abbreviated, change them, limit abbreviation to the `title` and not `description`, or something else?

henrikbrixandersen commented:
Beware that titles can be used for more than listing board features. When searching, for instance, it's really nice to have the vendor name, not just the vendor prefix, along with the name of the peripheral. 

Also, I would encourage staying as close to the naming in the vendor datasheet as possible for the title, as this is what users will be searching for (this goes both ways; searching the datasheet from a binding - or searching for a binding after reading the datasheet).

asmellby commented:
Made some modifications to this PR using my best judgment, absent additional input:

* Modified the implementation to only attempt auto-abbreviation on descriptions sourced from the `title` property
* Added the vendor name back to the titles

Example of how this looks now:
![Screenshot 2025-04-30 at 12 30 17](https://github.com/user-attachments/assets/ccab2428-20ab-41e7-8470-0b3a60b633ee)

asmellby committed:
doc: bindings-syntax: Document requirements for title property

Add documentation for the expectations for the title property
in devicetree bindings.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby commented:
Added some text to the bindings documentation spelling out the guidelines that have been raised so far. Also added the rules from dt-schema (max 100 characters, don't use the word "binding") to maximize compatibility.

kartben commented on doc/build/dts/bindings-syntax.rst:
```diff
@@ -73,8 +73,18 @@ These keys are explained in the following sections.
+vendor datasheet as possible.
+
+Titles should not exceed 100 characters. The description field should be used
+for longer descriptions. The words "binding" or "schema" should not be used
+in the title, everything is a binding.
```
suggesting to also add "driver" as a discouraged term. In fact, we should also discourage the usage of "driver" in descriptions as well. (see https://docs.kernel.org/devicetree/bindings/writing-bindings.html#:~:text=DON%E2%80%99T%20refer%20to%20Linux%20or%20%E2%80%9Cdevice%20driver%E2%80%9D%20in%20bindings.%20Bindings%20should%20be%20based%20on%20what%20the%20hardware%20has%2C%20not%20what%20an%20OS%20and%20driver%20currently%20support.)

```suggestion
for longer descriptions. The words "binding", "schema", or "driver" should not be used
in the title, everything is a binding.
```

asmellby replied:
Done!

kartben reviewed (changes requested):
very very nice - thanks again! Really just a minor request for change.
